js/eventloop/eventloop.go

```diff
@@ -55,11 +55,69 @@ func (e *EventLoop) wakeup() {
+// some asynchronous work off the main thread and that you may need to execute
+// some code back on the main thread when you are done. So, once you call this
+// method, the event loop will wait for you to finish and give it the callback
+// it needs to run back on the main thread before it can end the whole current
+// script iteration.
+//
+// ReservePendingCallback() *must* be called from the main runtime thread, but
+// its result scheduleOnMainThread() is thread-safe and can be called from any
+// goroutine. scheduleOnMainThread() ensures that its callback parameter is
+// added to the VU runtime's tasks queue, to be executed on the main runtime
+// thread eventually, when the VU is done with the other tasks before it. Unless
+// the whole event loop has been stopped, invoking scheduleOnMainThread() will
+// queue its argument and "wake up" the loop (if it was idle, but not stopped).
+//
+// Keep in mind that once you call ReservePendingCallback(), you *must* also
+// call scheduleOnMainThread() exactly once, even if don't actually need to run
+// any code on the main thread. If that's the case, you can pass an empty no-op
+// callback to it, but you must call it! The event loop will wait for the
+// scheduleOnMainThread() invocation and the k6 iteration won't finish and will
+// be stuck until the VU itself has been stopped (e.g. because the whole test or
+// scenario has ended). Any error returned by any callback on the main thread
+// will abort the current iteration and no further event loop callbacks will be
+// executed this iteration.
+//
+// A common pattern for async work is something like this:
+//
+//    func doAsyncWork(vu modules.VU) *goja.Promise {
+//        scheduleOnMainThread := vu.ReservePendingCallback()
+//        p, resolve, reject := vu.Runtime().NewPromise()
+//
+//        // Do the actual async work in a new independent goroutine, but make
+//        // sure that the Promise resolution is done on the main thread:
+//        go func() {
+//            // Also make sure to abort early if the context is cancelled, so
+//            // the VU is not stuck when the scenario ends or Ctrl+C is used:
+//            result, err := doTheActualAsyncWork(vu.Context())
+//            scheduleOnMainThread(func() error {
+//                if err != nil {
+//                    reject(err)
+//                } else {
+//                    resolve(result)
+//                }
+//                return nil  // do not abort the iteration
+//            })
+//        }()
+//
+//        return p
+//    }
+//
+// This ensures that the actual work happens asynchronously, while the Promise
+// is immediately returned and the main thread resumes execution. It also
+// ensures that the Promise resolution happens safely back on the main thread
+// once the async work is done, as required by goja and all other JS runtimes.
+func (e *EventLoop) ReservePendingCallback() (scheduleOnMainThread func(callback func() error)) {
```

Comment:
To be honest, the new name doesn't look better :man_shrugging: 

Maybe, or something similar.
```suggestion
func (e *EventLoop) RegisterCallbackQuota() (releaseQuota func(callback func() error)) {
```

Anyway, whatever the name will be at the end, please don't forget to fix the text in a panic below where now it stays `RegisterCallback called twice`.

Reply:
I'll fix all `RegisterCallback` mentions except the deprecated method when we reach consensus about the new name. I have a question about your name suggestion though - why ["quota"](https://www.merriam-webster.com/dictionary/quota)? :confused:

Reply:
@na-- personally, for me ["quota"](https://dictionary.cambridge.org/dictionary/english/quota) it represents what exactly this method is about:

> a fixed limit on the amount of something that someone is allowed to have or is expected to do

But as I said, I'm not considering myself the best person to define the new name. The only thing is that the current suggestion is still not ideal since, from the name of the method, it's hard to understand what the method is about.

Reply:
even by the "a fixed, limited amount or number that is officially allowed" definition, quota definitely doesn't represent what is going on here - we're dealing with neither a "fixed" nor a "limited" number of slots, nor is there any restriction to officially "allow" anything :confused: calling this function reserves you a single execution slot, but you can call it an unlimited number of times, there is no upper limit set by the VU, only the scenario/test duration limits it

so hard :-1: for "quota" from me, it adds more confusion than clarity

Reply:
As I said, I'm fine with that. But on the other note please count this comment as :-1: for the suggested `ReservePendingCallback`.

Reply:
And to be more constructive, I'm joining this suggestion https://github.com/grafana/k6/pull/2615#issuecomment-1192798652

Reply:
Can you explain why? 

I already replaced `Register` with `Reserve`, since I also thought that was slightly betterm but as I said in https://github.com/grafana/k6/pull/2615#issuecomment-1193653699, `Pending` also caries useful information, even more so than `Reserve` or `Register`... It's probably the most essential part of the whole name, if we go by its [definitions](https://dictionary.cambridge.org/dictionary/english/pending): "about to happen or waiting to happen", "used to say that one thing must wait until another thing happens", "waiting to happen", "while waiting for something to happen or be decided", i.e. exactly what we want to communicate here...

Reply:
Sure. Simply because I find the "pending" part is overthinking. From `Reserve` it's already clear that callback won't be executed immediately, so it already carries the information.